in-progress/8131-forced-inclusion.md

```diff
@@ -0,0 +1,522 @@
+  super.submit_next_epoch_proof(proof, tx_count, archive, fees)
+  forced_until = self.force_until(epoch, tx_count)
+
+  fip_block = 0
+  fip_index = 0
+  for fip in fips:
+    assert fip.block_number > last_block_number_in_prior_epoch
+    assert fip.block_number <= last_block_number_in_current_epoch
+
+    if fip.block_number != fip_block:
+      fip_block = fip.block_number
+      fip_index = 0
+
+    assert fip.index_in_block == fip_index, 'incorrect ordering'
+    assert self.show_included(fip), 'did not progress the queue, incorrect ordering'
+
+    fip_index += 1
+
+  assert forced_until <= self.forced_inclusion_tip, 'missing force inclusions'
+```
+
+The reason we have the constraints on the block being in the epoch is fairly simple.
+If we did not have this check, it would be possible to provide a proof for a transactions that was included in a previous epoch.
+This could happen for example if you also had a block in the prior epoch.
+Lets say that there are 2 transactions that have passed their "deadline" so they must be included.
+You provide a block with 2 transactions, the first is from the forced queue, but the second is not.
+The `force_until` would return that you need to progress by 2.
+You then provide it the proof for the prior tx, it progress by 1, and then the proof for your new, the second.
+At this point, you progressed by 2, but only one of them was actually new.
+
+If someone have included a forced transaction earlier than it was required, anyone can use the `show_included` to mark it as included and progress the state of the delayed queue.
+
+Notice that the forced inclusions outlined does **not** strictly require insertion into the beginning of the epoch.
+But only that it is within the start of blocks within the epoch.
+
+Also, observe, that a failure to include the forced transactions will be caught at the time of proof verification, not at the time of proposal!
+
+#### Stricter inclusion constraints
+
+By being stricter with our constraints, and more loose with our gas expenditure 😎 we can change the system such that it:
+- Enforces the ordering earlier in the life-cycle;
+- Force the delayed queue to be included as the first transactions, e.g., from the very start of the epoch, limiting the proposers power.
+
+By extending the checks of the proposal, we can cause an early failure, such that we don't have to wait until the block have been proven to catch them meddling with the force inclusion ordering.
+
+It requires a bit more accounting, and will likely be slightly more expensive, however, the improved properties seems to make up for that.
+
+Could I not have written this in the first go?
+Well, yes, but it might get confusing with both the pending and non pending functions so I'm going to keep them separate for now.
+
+
+```python
+active_pending: uint256
+forced_pending_tip: HashMap[uint256, uint256]
+included_pending: HashMap[uint256, HashMap[uint256, bool]]
```

Comment:
Can you sketch what these are supposed to be tracking? I can see that we're now allowing force inclusion proofs per pending block, but I'm not understanding the structure/flow with these new tips.

Reply:
The tip is just tracking how far in the "chain of forced transactions" we are.